### PR TITLE
Create a CouchDB storage example using petsets

### PR DESCRIPTION
**Pull request by kocolosk**
Pre-review before heading upstream

### CODE REVIEW

File: examples/storage/couchdb/core-config-maps/cluster-adm.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cluster-adm
+  namespace: default
+type: Opaque
+data:
+  password: Y2hhbmdlbWU=
+  username: YWRtaW4=
```
Comment (mikebrow) [inline examples/storage/couchdb/core-config-maps/cluster-adm.yaml:9]: nit.. missing new line char at the end of many of the files.

File: examples/storage/couchdb/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+service "kubetest001" created
+```
+
+### Creating the PetSet
+Now that we have the necessary configmaps, secrates and the services created let's create the couchdb petset!
+```
+kubectl create -f couchdb-petset.yaml 
+petset "kubetest001" created
+```
+You can now checkout the kubernetes dash to see the newly create couchdb cluster.  If your running minikube you can discover the cluster address as follows:
```
Comment (mikebrow) [inline examples/storage/couchdb/README.md:39]: nit... "If your running” /s/your/you’re/

"You can now checkout the kubernetes dash to see the newly create couchdb cluster.” Dashboard? You mention it but then proceed to do something else unrelated to observe if the service is running.

The example creates a couchdb cluster but does not really exercise it in the context of a cluster.

If beyond the scope of this exercise maybe link to somewhere for the reader to follow for use of the cluster’d couchdb?

File: examples/storage/couchdb/core-config-maps/cluster-adm.yaml (hunk 1 above)
Comment (MHBauer) [reply examples/storage/couchdb/core-config-maps/cluster-adm.yaml:9]: agree

File: examples/storage/couchdb/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+There are two services we need to create for our couchdb cluster.  First theres the headless service for pod to pod discovery using DNS lookup and to give static hostname. Then we have the service that exposes the couchdb cluster. Let us create them:
+
+```
+kubectl create -f couchdb-petset-headless-service.yaml -f couchdb-petset-service.yaml 
+service "kubetest001-backend" created
+service "kubetest001" created
+```
+
+### Creating the PetSet
+Now that we have the necessary configmaps, secrates and the services created let's create the couchdb petset!
```
Comment (mikebrow) [inline examples/storage/couchdb/README.md:34]: /s/secrates/secrets/

File: examples/storage/couchdb/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+```
+kubectl create -f core-config-maps/core-configmap.yaml -f core-config-maps/cluster-adm.yaml 
+configmap "core-config" created
+secret "cluster-adm" created
+```
+#### Note on Admin users
+We have populated the `local.ini` config file with user "admin" and the plain text password "changeme". On startup couchdb will hash the users password.  If you would like to define your own admin users then just add them to local.ini section of the core-config.  Also make sure that there is a valid admin user credentials in the cluster admin secret at `core-config-maps/cluster-adm.yam` bash64 encoded, naturally. This is needed for the utility pods if misconfiger we are going to have a bad time.
```
Comment (mikebrow) [inline examples/storage/couchdb/README.md:18]: /s/misconfiger/misconfigured/

File: examples/storage/couchdb/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
+
+## Couchdb 2.0.0 on Kubernetes
+This document describes the steps to create a Couchdb 2.0.0 cluster on Kubernetes.
+
+### Prerequisites
+Access to a kubernetes cluster running v1.4 or later, kubectl installed and configured.  Minikube was used for this example. Note that the service type for the couchdb cluster is set to "NodePort" as the loadbalancer type is unsupported by minikube.
```
Comment (mikebrow) [inline examples/storage/couchdb/README.md:6]: pls use Kube's documentation standard for line wrapping at 80col and single space for sentences.

Comment (MHBauer) [reply examples/storage/couchdb/README.md:6]: it makes it easier to comment on specific items in a line.

File: examples/storage/couchdb/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+### Prerequisites
+Access to a kubernetes cluster running v1.4 or later, kubectl installed and configured.  Minikube was used for this example. Note that the service type for the couchdb cluster is set to "NodePort" as the loadbalancer type is unsupported by minikube. 
+ 
+
+### Creating the Configmaps and secrets 
+
+Couchdb has erlang at its core and as such needs to be correctly configured to allow for clustering within Kubernetes.  There are currently three config files `vm.args`, `local.ini` and `default.ini` that are stored as blobs in a [ConfigMap](http://kubernetes.io/docs/user-guide/configmap/) found at `core-config-maps/core-configmap.yaml`. The `vm.args` file is where we set the erlang 'magic' cookie for all the couchdb instances alowing them to communicate with each other.  Before we get to creating our couchdb cluster these config maps need to be created:
+```
+kubectl create -f core-config-maps/core-configmap.yaml -f core-config-maps/cluster-adm.yaml
```
Comment (MHBauer) [inline examples/storage/couchdb/README.md:13]: add a `$` at the beginning of the line to indicate a shell command.

File: examples/storage/couchdb/README.md (hunk 4 above)
Comment (MHBauer) [reply examples/storage/couchdb/README.md:18]: I appreciate the folksy-ness with 'bash64 encoded, naturally', but these are instructions, so I'd prefer if it was directly stated exactly what needed to be done.

**Comment by MikeSpreitzer** [comment]
The master branch here is intended to be a copy of the master branch from https://github.com/kubernetes/kubernetes, no independent merges here.  There should be no PRs against this branch.  If you want to share work with a limited circle of reviewers before heading upstream then a different mechanism should be used.  Would email referencing a feature branch work?

File: examples/storage/couchdb/README.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+```
+#### Note on Admin users
+We have populated the `local.ini` config file with user "admin" and the plain text password "changeme". On startup couchdb will hash the users password.  If you would like to define your own admin users then just add them to local.ini section of the core-config.  Also make sure that there is a valid admin user credentials in the cluster admin secret at `core-config-maps/cluster-adm.yam` bash64 encoded, naturally. This is needed for the utility pods if misconfiger we are going to have a bad time. 
+
+
+### Creating the services 
+
+Along with the erlang cookie, Couchdb instances need to have a static network identity.  This static identity is used to set the erlang node names for the couchdb instances.  To achieve this in Kubernetes petsets are leveraged giving every couchdb pod its own DNS name for its lifetime. 
+  
+There are two services we need to create for our couchdb cluster.  First theres the headless service for pod to pod discovery using DNS lookup and to give static hostname. Then we have the service that exposes the couchdb cluster. Let us create them:
```
Comment (MHBauer) [inline examples/storage/couchdb/README.md:25]: suggestion, `The first service we create is for pod to pod` & `The second service we create exposes the couchdb`

I'm not sure what the 'headless' service means in this case. Can we drop that or is it important?

**Comment by kocolosk** [comment]
@MikeSpreitzer that won't allow for the same level of inline commenting and code review. FWIW no one intends to merge this PR; it's purely to review before submitting upstream. Master will stay clean. The alternative was to have @James1912 submit a PR to himself - this seemed a better way to provide visibility to the discussion.

**Comment by MHBauer** [comment]
so I can leave my comments. I'll close after.

**Review by MHBauer — COMMENTED** [review]
comments

**Comment by kocolosk** [comment]
How about we just leave it open and promise not to hit the merge button? That way @James1912 can add more commits to address the reviewers comments. What do you think @MikeSpreitzer ?

File: examples/storage/couchdb/README.md (hunk 4 above)
Comment (James1912) [reply examples/storage/couchdb/README.md:18]: Good pint. Would pointing the reader at the doc's be sufficient here? All "manually" created secrets have to be base64 encoded so I thought I would not reiterate that here. 

http://kubernetes.io/docs/user-guide/secrets/

File: examples/storage/couchdb/README.md (hunk 2 above)
Comment (James1912) [reply examples/storage/couchdb/README.md:39]: Your right, given that I already reference minikube cmd's for getting the service IP I could also add in there the step to get up the k8s dash.  A `minikube dashboard` here would not go amiss.  What I did not want to do is bog people who are not using minikube with minikube guff. Also I could point them to the couchdb docs I suppose 😄

**Comment by James1912** [comment]
Thanks for the comments @mikebrow, @MHBauer.  Appreciate your time looking through this

**Comment by MHBauer** [comment]
@James1912 we have more. Can you open up a PR in your own local repo and we'll do this discussion over there?

**Comment by James1912** [comment]
Sure thing: https://github.com/James1912/kubernetes/pull/1

File: examples/storage/couchdb/README.md (hunk 7 above)
Comment (James1912) [reply examples/storage/couchdb/README.md:25]: Its a service that does not act as a load balancer.  In this case the headless service is there to give every pod in the petset a DNS entry: 

http://kubernetes.io/docs/user-guide/services/#headless-services

Comment (MHBauer) [reply examples/storage/couchdb/README.md:25]: thanks!

File: examples/storage/couchdb/README.md (hunk 4 above)
Comment (MHBauer) [reply examples/storage/couchdb/README.md:18]: Ah, okay. A doc link is always good in my opinion. I'd err on too much info rather than too little.

**Pull request closed without merging**